logstash-core/src/main/java/org/logstash/plugins/discovery/PluginRegistry.java

```diff
@@ -37,40 +37,58 @@
 import java.util.Set;
 
 /**
- * Registry for built-in Java plugins (not installed via logstash-plugin)
- */
+ * Registry for built-in Java plugins (not installed via logstash-plugin).
+ * This is singleton ofr two reasons:
+ * <ul>
+ *  <li>it's a registry so no need for multiple instances</li>
+ *  <li>the Reflections library used need to run in single thread during discovery phase</li>
+ * </ul>
+ * */
 public final class PluginRegistry {
 
-    private static final Map<String, Class<Input>> INPUTS = new HashMap<>();
-    private static final Map<String, Class<Filter>> FILTERS = new HashMap<>();
-    private static final Map<String, Class<Output>> OUTPUTS = new HashMap<>();
-    private static final Map<String, Class<Codec>> CODECS = new HashMap<>();
+    private final Map<String, Class<Input>> inputs = new HashMap<>();
+    private final Map<String, Class<Filter>> filters = new HashMap<>();
+    private final Map<String, Class<Output>> outputs = new HashMap<>();
+    private final Map<String, Class<Codec>> codecs = new HashMap<>();
+    private static final Object lock = new Object();
+    private static PluginRegistry instance;
 
-    static {
+    private PluginRegistry() {
         discoverPlugins();
     }
 
-    private PluginRegistry() {} // utility class
-
+    public static PluginRegistry getInstance() {
```

Comment:
do we really need the lazy initialization here? (cause I think it should be marked `volatile` otherwise)
couldn't we just have a `private static final PluginRegistry INSTANCE = new PluginRegistry()` ?
(the class will initialize on it's first use, or if that is an issue let's have a nested static class to hold the `INSTANCE`)

Reply:
The problem with `private static final PluginRegistry INSTANCE = new PluginRegistry()` is that statement is executed when the class is loaded, and is really similar to have a `static` block that does all the initialization. The problem with this, is that we don't have direct control of "when" the PluginRegistry is instantiated, but the order is controlled by import/require sequence.
Having this control could be helpful in writing tests for the alias part (I hope).

Regarding the `volatile`, if we used it we could potentially have many threads (when we have multiple pipelines) that creates more instances, but then only one is saved and used by the other.
Creating multiple instances could raise some problems with the `reflections` library, because these 2 lines:
```java
Reflections reflections = new Reflections("org.logstash.plugins");
Set<Class<?>> annotated = reflections.getTypesAnnotatedWith(LogstashPlugin.class);
```

must be executed in single thread, and previously this condition was granted by the `static {}` initializer that's executed by the class loading mechanism and that's granted to be single thread, else we risk problems that one threads executing `reflections` closes the jar file to the other thread that's executing the classpath scan. It could raise problems like:
```
[DEBUG] 2021-04-08 09:21:08.935 [Agent thread] agent - Converging pipelines state {:actions_count=>2}
[DEBUG] 2021-04-08 09:21:08.944 [Converge PipelineAction::Create<test>] agent - Executing action {:action=>LogStash::PipelineAction::Create/pipeline_id:test}
[DEBUG] 2021-04-08 09:21:08.944 [Converge PipelineAction::Create<test2>] agent - Executing action {:action=>LogStash::PipelineAction::Create/pipeline_id:test2}
[INFO ] 2021-04-08 09:21:09.094 [Api Webserver] agent - Successfully started Logstash API endpoint {:port=>9600}
[DEBUG] 2021-04-08 09:21:09.358 [Converge PipelineAction::Create<test>] Reflections - going to scan these urls:
jar:file:/home/andrea/workspace/logstash_andsel/build/logstash-8.0.0-SNAPSHOT/logstash-core/lib/jars/logstash-core.jar!/
[DEBUG] 2021-04-08 09:21:09.358 [Converge PipelineAction::Create<test2>] Reflections - going to scan these urls:
jar:file:/home/andrea/workspace/logstash_andsel/build/logstash-8.0.0-SNAPSHOT/logstash-core/lib/jars/logstash-core.jar!/
[INFO ] 2021-04-08 09:21:09.395 [Converge PipelineAction::Create<test>] Reflections - Reflections took 35 ms to scan 1 urls, producing 24 keys and 48 values 
[ERROR] 2021-04-08 09:21:09.397 [Converge PipelineAction::Create<test2>] agent - Failed to execute action {
	:action=>LogStash::PipelineAction::Create/pipeline_id:test2, 
	:exception=>"Java::JavaLang::IllegalStateException", 
	:message=>"zip file closed", 
	:backtrace=>["java.base/java.util.zip.ZipFile.ensureOpen(ZipFile.java:915)",
                 "java.base/java.util.zip.ZipFile$ZipEntryIterator.next(ZipFile.java:525)",
                 "java.base/java.util.zip.ZipFile$ZipEntryIterator.nextElement(ZipFile.java:518)",
                 "java.base/java.util.zip.ZipFile$ZipEntryIterator.nextElement(ZipFile.java:494)",
                 "org.reflections.vfs.ZipDir$1$1.computeNext(ZipDir.java:31)",
                 "org.reflections.vfs.ZipDir$1$1.computeNext(ZipDir.java:26)",
                 "com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:141)",
                 "com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:136)",
                 "org.reflections.Reflections.scan(Reflections.java:243)",
                 "org.reflections.Reflections.scan(Reflections.java:202)",
                 "org.reflections.Reflections.<init>(Reflections.java:123)",
                 "org.reflections.Reflections.<init>(Reflections.java:168)",
                 "org.reflections.Reflections.<init>(Reflections.java:141)",
                 "org.logstash.plugins.discovery.PluginRegistry.discoverPlugins(PluginRegistry.java:55)",
                 "org.logstash.plugins.discovery.PluginRegistry.<init>(PluginRegistry.java:50)",
                 "org.logstash.plugins.factory.PluginFactoryExt.<init>(PluginFactoryExt.java:85)",
                 "org.logstash.execution.JavaBasePipelineExt.initialize(JavaBasePipelineExt.java:73)",
                 "org.logstash.execution.JavaBasePipelineExt$INVOKER$i$1$0$initialize.call(JavaBasePipelineExt$INVOKER$i$1$0$initialize.gen)",
                 "org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:837)",
                 "org.jruby.ir.runtime.IRRuntimeHelpers.instanceSuper(IRRuntimeHelpers.java:1169)",
                 "org.jruby.ir.runtime.IRRuntimeHelpers.instanceSuperSplatArgs(IRRuntimeHelpers.java:1156)",
                 "org.jruby.ir.targets.InstanceSuperInvokeSite.invoke(InstanceSuperInvokeSite.java:39)",
                 "home.andrea.workspace.logstash_andsel.build.logstash_minus_8_dot_0_dot_0_minus_SNAPSHOT.logstash_minus_core.lib.logstash.java_pipeline.RUBY$method$initialize$0(/home/andrea/workspace/logstash_andsel/build/logstash-8.0.0-SNAPSHOT/logstash-core/lib/logstash/java_pipeline.rb:47)",
                 "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:80)",
                 "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:70)",
                 "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:332)",
                 "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:86)",
                 "org.jruby.RubyClass.newInstance(RubyClass.java:939)",
                 "org.jruby.RubyClass$INVOKER$i$newInstance.call(RubyClass$INVOKER$i$newInstance.gen)",
                 "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:332)",
                 "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:86)",
                 "org.jruby.ir.instructions.CallBase.interpret(CallBase.java:549)",
                 "org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:361)",
                 "org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)",
                 "org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:86)",
                 "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:73)",
                 "org.jruby.ir.targets.InvokeSite.invoke(InvokeSite.java:207)",
                 "home.andrea.workspace.logstash_andsel.build.logstash_minus_8_dot_0_dot_0_minus_SNAPSHOT.logstash_minus_core.lib.logstash.agent.RUBY$block$converge_state$2(/home/andrea/workspace/logstash_andsel/build/logstash-8.0.0-SNAPSHOT/logstash-core/lib/logstash/agent.rb:382)",
                 "org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:138)",
                 "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)",
                 "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)",
                 "org.jruby.runtime.Block.call(Block.java:139)",
                 "org.jruby.RubyProc.call(RubyProc.java:318)",
                 "org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:105)",
                 "java.base/java.lang.Thread.run(Thread.java:834)"]}
```

Reply:
OK, thanks for the explanation. 

> (cause I think it should be marked `volatile` otherwise)

What I meant is to mark the static instance to prevent issues like "cache incoherence" (some [read](https://dzone.com/articles/demystifying-volatile-and-synchronized-again))

```java
private static volatile PluginRegistry instance;

    public static PluginRegistry getInstance() {
        PluginRegistry instance = PluginRegistry.instance;
        if (instance == null) {
            synchronized (PluginRegistry.class) {
                instance = PluginRegistry.instance;
                if (instance == null) {
                    instance = PluginRegistry.instance = new PluginRegistry();
                }
            }
        }
        return instance;
    }
```